File: fixtures/sample-repos/selenium/pages/actions/login_actions.py
"""Action class: receives the locator class, owns user workflows.

Spec files own assertions; this module never calls `assert`.
"""
from selenium.webdriver.remote.webdriver import WebDriver

from pages.locators.login_locators import LoginLocators


class LoginActions:
    def __init__(self, driver: WebDriver):
        self.locators = LoginLocators(driver)

    def navigate_to(self, path: str) -> None:
        self.locators.driver.get(f"http://localhost:3000{path}")

    def sign_in(self, email: str, password: str) -> None:
        email_el = self.locators.visible(self.locators.EMAIL)
        password_el = self.locators.visible(self.locators.PASSWORD)
        email_el.clear()
        email_el.send_keys(email)
        password_el.clear()
        password_el.send_keys(password)
        self.locators.visible(self.locators.SIGN_IN).click()

File: fixtures/sample-repos/webdriverio/test/pageobjects/actions/LoginActions.ts
import { LoginLocators } from '../locators/LoginLocators';

// Action class: receives the locator class, owns user workflows.
// Specs own assertions; actions return state or perform interactions.
export class LoginActions {
  private locators = new LoginLocators();

  async navigateTo(path: string): Promise<void> {
    await browser.url(path);
  }

  async signIn(email: string, password: string): Promise<void> {
    await this.locators.emailInput.waitForDisplayed();
    await this.locators.emailInput.setValue(email);
    await this.locators.passwordInput.setValue(password);
    await this.locators.signInButton.click();
  }
}
